### ISSUE TITLE
missing comptime undefined propagation for equality operator with pointer, optional, error union, and integers

### ISSUE THREAD

**Issue by Jarred-Sumner** [ziglang/zig#17798]
The following strictly always is undefined behavior that Zig chooses to compile successfully:

Checking if a value is not equal to undefined:
```zig
pub fn deinit(this: *@This()) void {
   if (this.context != undefined) {
    this.allocator.destroy(this.context);
  }
}
```
Checking if a value is equal to undefined:
```zig
pub fn deinit(this: *@This()) void {
   if (this.context == undefined) {
    this.allocator.destroy(this.context);
  }
}
```

Why isn't this a compiler error?

**Comment by Jarred-Sumner** [comment]
Similarly, this test passes in a debug build. It should be a compiler error?

```zig
test {
    var foo: ?*bar.baz.Foo = undefined;
    if (foo != null and foo != undefined) {
        @import("std").debug.print("wat", .{});
    }
}
```

**Comment by andrewrk** [comment]
```zig
test "undefined propagation in equality operators for bool" {
    var foo: bool = undefined;
    if (foo == undefined) {}
}

test "undefined propagation in equality operators for pointer" {
    var foo: *i32 = undefined;
    if (foo == undefined) {}
}

test "undefined propagation in equality operators for optional" {
    var foo: ?*i32 = undefined;
    if (foo == undefined) {}
}

test "undefined propagation in equality operators for error union" {
    var foo: anyerror = undefined;
    if (foo == @as(anyerror!i32, undefined)) {}
}

test "undefined propagation in equality operators for integers" {
    var foo: i32 = undefined;
    if (foo == undefined) {}
}
```

Observed output:

```
test.zig:3:13: error: use of undefined value here causes undefined behavior
    if (foo == undefined) {}
        ~~~~^~~~~~~~~~~~
```

Expected output:

The boolean one is correct. The compile error is missing for all the other cases.

**Comment by andrewrk** [comment]
Fix reverted in 70d8baaec11ca370b73fce72d7f3dfce2277455b.
